File: [0121][Easy][Best_Time_to_Buy_and_Sell_Stock]/Best_Time_to_Buy_and_Sell_Stock.py
# solution 2, one time traversal of the price list.
# leetcode time     cost : 56 ms
# leetcode memory   cost : 12.5 MB 
# Time  Complexity: O(N)
# Space Complexity: O(1)
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_profit, min_price = 0, float('inf')
        for price in prices:
            min_price = min(min_price, price)
            profit = price - min_price
            max_profit = max(max_profit, profit)
        return max_profit

File: [0121][Easy][Best_Time_to_Buy_and_Sell_Stock]/Best_Time_to_Buy_and_Sell_Stock.java
// solution 1, brute force calculation.
// leetcode time     cost : 318 ms
// leetcode memory   cost : 38.5 MB 
// Time  Complexity: O(N*N)
// Space Complexity: O(1)
public class Solution {
    public int maxProfit(int prices[]) {
        int maxprofit = 0;
        for (int i = 0; i < prices.length - 1; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                int profit = prices[j] - prices[i];
                if (profit > maxprofit)
                    maxprofit = profit;
            }
        }
        return maxprofit;
    }
}